src/Cli/func/Program.cs

```diff
@@ -32,14 +34,26 @@ internal static void Main(string[] args)
             SetupGlobalExceptionHandler();
             SetCoreToolsEnvironmentVariables(args);
             _container = InitializeAutofacContainer();
+            var processManager = _container.Resolve<IProcessManager>();
             AppDomain.CurrentDomain.ProcessExit += CurrentDomain_ProcessExit;
 
-            Console.CancelKeyPress += (s, e) =>
-            {
-                _container.Resolve<IProcessManager>()?.KillChildProcesses();
-            };
+            CancelKeyHandler.Register(
+                processManager: processManager,
+                onShuttingDown: _shuttingDownCts.Cancel,
+                onGracePeriodTimeout: _forceShutdownCts.Cancel);
 
-            ConsoleApp.Run<Program>(args, _container);
+            try
+            {
+                await ConsoleApp.RunAsync<Program>(args, _container, _shuttingDownCts.Token).WaitAsync(_forceShutdownCts.Token);
+            }
+            catch (OperationCanceledException)
```

Comment:
We may only want to call `processManager` when it is specifically the `_forceShutdownCts` that is cancelled.

Consider:


```suggestion
            catch (OperationCanceledException ex)
            {
                if (ex.CancellationToken == _forceShutdownCts.Token)
                {
                    // processManager stuff
                }
            }
```